### ISSUE TITLE
Call tinyMCE.triggerSave() before sending metabox data

### ISSUE THREAD

**Issue by danielbachhuber** [WordPress/gutenberg#7176]
Prior to sending metabox data for save, we need to call `window.tinyMCE.triggerSave()` to update any `<textarea>` fields controlled by `wp_editor()` instances.

If TinyMCE is used on a `<textarea>` field, then the data doesn't end up in the actual `<textarea>` until a save is triggered.

Originally #6293

**Comment by krisgale** [comment]
workaround:

```
document.addEventListener( 'DOMContentLoaded', () => {
  wp.editor.initialize( 'field_name', {
    'textarea_name': 'field_name',
    'tinymce': {
      'init_instance_callback': ( editor ) => {
        editor.on( 'change', function() {
          const editor = tinymce.get( this.id );
          document.getElementById( 'field_name' ).value = editor.getContent();
        } );
      },
    },
  } );
} );
```

where `field_name` is the `id=` and `name=` of the `textarea` being 'enhanced.'

**Comment by danielbachhuber** [comment]
@krisgale Can you share a code snippet and testing instructions to reproduce your specific problem? I'd like to verify the change in #8762

**Comment by krisgale** [comment]
no code snippet as such, but adding any metabox that contains an instance of `wp_editor` would replicate the issue...  disable gutenberg, content saves; enable gutenberg, content does not save.

**Comment by danielbachhuber** [comment]
I put together a code snippet:

```
add_action( 'add_meta_boxes', function(){
	add_meta_box( 'test_tinymce', 'Test TinyMCE', function( $post ){
		 $field_value = get_post_meta( $post->ID, 'test_tinymce', true );
		 wp_editor( $field_value, 'test_tinymce_id', array(
			'wpautop'       => true,
			'media_buttons' => false,
			'textarea_name' => 'test_tinymce',
			'textarea_rows' => 10,
			'teeny'         => true
		) );
	}, null, 'advanced', 'high' );
});
add_action( 'save_post', function( $post_id ){
	if ( ! isset( $_POST['test_tinymce'] ) ) {
		return;
	}
	update_post_meta( $post_id, 'test_tinymce', $_POST['test_tinymce'] );
});
```

**Comment by krisgale** [comment]
@danielbachhuber thanks for staying on top of it

**Comment by michakrapp** [comment]
So the current state is that Gutenberg will not save content from `wp_editor()` fields in metaboxes, correct?

@krisgale Unfortunately I can't get your workaround to correctly work.
`field_name` should be the `$editor_id` parameter you set on `wp_editor()` corrrect? 
So my code would look like this:
```
document.addEventListener( 'DOMContentLoaded', () => {
    wp.editor.initialize( '_wp_editor_eve_post_links', {
        'textarea_name': '_wp_editor_eve_post_links',
        'tinymce': {
            'init_instance_callback': ( editor ) => {
                editor.on( 'change', function() {
                    const editor = tinymce.get( this.id );
                    document.getElementById( '_wp_editor_eve_post_links' ).value = editor.getContent();
                } );
            },
        },
    } );
} );
```

But the `change` event does not get triggered.

**Comment by krisgale** [comment]
important note per the above ^
in implementing this workaround, i *replaced* the call in php to `wp_editor()` with only js to instantiate the tinymce instance

**Comment by michakrapp** [comment]
@danielbachhuber has this bug been fixed with a change in Version 3.9 or 3.8? (I think with 3.7 the issue has existed)

Because after the newest update it seams to work now.

**Comment by danielbachhuber** [comment]
@michakrapp I'm not sure whether the issue has been fixed accidentally. When we worked on this last, we weren't able to reproduce the issue reliably: https://github.com/WordPress/gutenberg/pull/8762#issuecomment-412137005

**Comment by michakrapp** [comment]
Strange thing - after the update it has been working a few times.
Now the error is back.

But I can confirm that call `window.tinyMCE.triggerSave()` before saving works like you demonstrated in [#8762](https://github.com/WordPress/gutenberg/pull/8762#issuecomment-412137005)

**Comment by ideadude** [comment]
I am able to recreate this issue with our memberlite-elements plugin. It adds a few wp_editor instances into the meta boxes below the page. When I click to update the page, the meta box fields aren't saved. When I debugged, it's because the $_POST['field_name'] value was not updated. It seems that TinyMCE would update this before the post got saved and the suggestion solution here would work.

I tried typing `window.tinyMCE.triggerSave()` into the JS console and then saving and that doesn't work for me. Something else is needed.

I am using WP 5.0-beta3-43878, Gutenberg Version 4.2.0, Chrome

FWIW here is where we insert our editor:
https://github.com/strangerstudios/memberlite-elements/blob/development/elements/page_banners.php#L73

Here is where we process it. It fails on this line since the POST content is empty.
https://github.com/strangerstudios/memberlite-elements/blob/development/elements/page_banners.php#L184

Note that this plugin is meant to work with the Memberlite theme, but does still work (in terms of saving values) with Twenty* themes and fails with a Twenty* theme active.

Do we need a better minimal example? Do we need someone to dig into this further and look for the fix? Let me know. I might have time.

**Comment by robincornett** [comment]
Wondering if this will be fixed in WP Core. Currently, any meta boxes with wysiwyg editors don't save in the block editor, and the promised [meta box compatibility flags](https://make.wordpress.org/core/2018/11/07/meta-box-compatibility-flags/) do not work either: meta boxes marked as incompatible are simply skipped for output, rather than falling back to the classic editor.

**Comment by pbrocks** [comment]
@robincornett I suspect that this will be fixed in WP Core when the issue is understood fully. In my view this is related to TinyMCE not loading correctly for Metaboxes. That is a developer concern. For users, however, it may be helpful to know that Metaboxes will save correctly and fully if done using the Text tab, not Visual. That is, if you make all of the changes that you wish using the Visual tab, ie using the TinyMCE editor, and you switch to the Text tab, meaning you will see the text rendering of the html markup, when you click Publish or Update, the data in the metaboxes, including the markup, will save. Because of I am confident that this will be corrected in Core, but it will take someone who understands the waterfall of javascript in order to place the `.triggerSave()` where it will be recognized and the change effectuated.

**Comment by robincornett** [comment]
Thanks, @pbrocks--I get that it will theoretically be fixed in Core, but at the moment, I'm blocked on updating a plugin because of this issue (telling users to use the text tab isn't a great option for me) combined with no backwards compatibility in place yet. I hate to tell my users that they have to add the Classic Editor plugin just to keep using the plugin they've been using for an extended period of time, but right now that is all I can do. My concern is as a developer, not as a user.

**Comment by webmandesign** [comment]
Having the same issue here. Can we hope this will get fixed for WordPress 5.0?

I can't imagine telling users to make changes in "Visual" tab, while switching to "Text" tab before saving/updating post so their changes are properly saved... That's is a very bad UX.

**Comment by mrhead** [comment]
According to the assigned milestone, it doesn't look that it's going to be fixed in WordPress 5.0 nor 5.0.1 which is a bummer.

**Comment by youknowriad** [comment]
I'm still interested in fixing this but last time I tried, it was not reproducible consistently which made the fix useless. If someone could provide me with a minimal metabox with the issue, I'd be happy to try to fix it again.

**Comment by robincornett** [comment]
I've just reconfirmed that this is still an issue, using:
* @danielbachhuber [sample metabox code](https://github.com/WordPress/gutenberg/issues/7176#issuecomment-412134455) (see above)
* WordPress 5.0-RC1-43944
* Twenty Nineteen

With no plugins active, the wp_editor content in the metabox does not save. With Gutenberg 4.5.1 active, the content does not save.

Note that to actually see the content failing to save, you need to reload the editor page, since Gutenberg is saving via Ajax. Until the page is reloaded, the content will appear in the editor in the metabox. The row is created in the post meta table, but with no value, and checking the `$_POST` array on save shows that the `test_tinymce` key is present, but empty.

**Comment by youknowriad** [comment]
Thanks I've used @danielbachhuber 's code previously and it was inconsistent, I'll try again very soon.

**Comment by webmandesign** [comment]
@youknowriad Thanks for looking into this issue! Here is something that might help reproducing it:

## Test environment:

I've put together a test theme you can download from https://github.com/webmandesign/_test/archive/test/gutenberg/metabox-wysiwyg.zip

You will also need to install and activate [WebMan Amplifier plugin](https://wordpress.org/plugins/webman-amplifier/). Just in case, please grab the latest plugin code from https://github.com/webmandesign/webman-amplifier/archive/master.zip

## How to test:

Once you install and activate the theme and WebMan Amplifier plugin, a new metabox is displayed on post/page edit screen, with a single "WYSIWYG test" field. To reproduce the issue here is my workflow:

1. Editing the "Hello world!" post, insert a "Lorem ipsum" text into the metabox's "WYSIWYG test" field's "Visual" tab. (And make the "ipsum" bold.)
2. Update the post and preview it: **[The issue:]** All you should see now is empty blue box on top of post content.
3. Back to post editing, refresh/reload the page: **[The issue:]** The content of metabox's "WYSIWYG test" field is empty.
4. Again, put "Lorem ipsum" text into the metabox's "WYSIWYG test" field's "Visual" tab. (And make the "ipsum" bold.)
5. Switch over to "Text" tab and update the post. Preview it: Now you should see our "Lorem ipsum" text within the blue box on top of post content.

## Additional info:

* All the time when you make changes to "WYSIWYG test" field's content in "Visual" tab, you need to switch to "Text" tab for the data to be properly saved before saving the actual post.
* When saving the post while still in "Visual" tab of "WYSIWYG test" field, the PHP `$_POST` for the `wm-wysiwyg` field ID is empty (or set to previous value). Only when you switch to "Text" tab the data are correctly updated and sent in `$_POST`.

Hope this will help to reproduce the issue.

**Comment by youknowriad** [comment]
Testing for the https://github.com/WordPress/gutenberg/pull/12363 PR would be highly appreciated.

**Comment by webmandesign** [comment]
@youknowriad Sorry, I don't really know how to compile Gutenberg code for usable WordPress installation/testing :( I always test with released versions only.

**Comment by youknowriad** [comment]
[gutenberg.zip](https://github.com/WordPress/gutenberg/files/2620619/gutenberg.zip)


Here's a zip if you want to test.

**Comment by robincornett** [comment]
@youknowriad thank you, I've tested your fix with the sample metabox code and with my own custom code and the wp_editor instances are now properly saving (as long as Gutenberg is active).

Once your fix is added to Gutenberg, will it also be copied over to WordPress Core? I would love to know if it will make it into 5.0 as that will affect some plugin development changes. Thank you!

**Comment by youknowriad** [comment]
5.0 is frozen at the moment. This will more likely land in 5.0.1 which should happen just two weeks after the 5.0 but this could change, not my decision.

**Comment by webmandesign** [comment]
Not sure if I'm doing anything wrong, but the fix doesn't seem to work for me (using the gutenberg.zip from above).

I've tested with WP4.9.8. Should I maybe test with WP5.0-rc1 and Gutenberg acitve?

BTW, I've also tested with Gutenberg disabled and in that case it works perfectly fine.

**Comment by youknowriad** [comment]
You should test with just the Gutenberg plugin active (from the zip). The WordPress version don't have any impact.

**Comment by robincornett** [comment]
I understand (about 5.0.1), although that's unfortunate with the current state of the incompatible metabox notices--this would mean I need to go ahead and update my plugin(s) with the metaboxes marked as incompatible with the block editor--so users will be told to install the Classic Editor. Even if this is fixed in 5.0.1, users who've encountered this will have activated the Classic Editor based on the warnings; I wonder how many will bother deactivating it, especially since the warning disappears in the Classic Editor once the metabox has been marked as compatible with the block editor.

It seems like affected users may enable the Classic Editor and then never know to leave it, even if they wanted to, once the compatibility warning disappears. I realize that the metabox notifications are a separate issue, but these two together are a concern to me.

**Comment by youknowriad** [comment]
@robincornett That's not the only thing that's not compatible with previous plugins with Gutenberg. And the solution has always been the classic editor until the plugin is made compatible. So I don't think it will have a big impact.

**Comment by webmandesign** [comment]
@youknowriad Sure, I must have written it badly ;) I've tested with Gutenberg (from the zip) active. And it wasn't working. Then, for comparison, I've deactivated Gutenberg and tested again, and it worked with WP4.9.8.
So, what's my point is that the fix from the zip file is not working for me.

**Comment by youknowriad** [comment]
@webmandesign Maybe it's fixing one incompatibility (the triggerSave call) and we're missing another one.

**Comment by youknowriad** [comment]
@azaozz Aside the `window.tinyMCE.triggerSave()` do you have anything that comes to your mind that may have an impact on the persistence of the TinyMCE fields?

**Comment by robincornett** [comment]
Just a note: the fix/PR works with Gutenberg active running 5.0RC, but not trunk (which I believe is still equivalent with 4.9.8, right?). So some difference between 5.0 and trunk is making the fix not work, it appears.

**Comment by webmandesign** [comment]
@youknowriad That may be possible.
I was actually digging into metabox builder code of my WebMan Amplifier plugin as I haven't touched that part for a long time. However:
* I [load TinyMCE editor](https://github.com/webmandesign/webman-amplifier/blob/master/includes/metabox/fields/texts.php#L251) in simple way.
* [I do not modify the TinyMCE via metabox JS](https://github.com/webmandesign/webman-amplifier/blob/master/assets/js/metabox.js) in any way.
* The [data saving](https://github.com/webmandesign/webman-amplifier/blob/master/includes/metabox/class-metabox.php#L632) seems OK to me (it works when not using Gutenberg).

So, I think we can rule out WebMan Amplifier issue here?

Maybe, is there any JS action triggered when Gutenberg post is being updated? So I can use it to hook `window.tinyMCE.triggerSave()` code from WebMan Amplifier to test?

**Comment by youknowriad** [comment]
@robincornett Interesting and also good news I'd say. Maybe it's related to the TinyMCE config alignments we did recently.

**Comment by youknowriad** [comment]
@webmandesign Could you test with 5.0 RC and this plugin to see if it fixes the issue?

**Comment by webmandesign** [comment]
@youknowriad Yes, I've done that immediately as I saw @robincornett 's comment. And I can confirm it works perfectly fine now and the issue is fixed! Great catch there @robincornett!

**Comment by youknowriad** [comment]
Awesome, this means we can merge the PR once approved and close this issue

**Comment by azaozz** [comment]
> Aside the window.tinyMCE.triggerSave() do you have anything that comes to your mind...

The other thing is that when the "default" `wp_editor()` is used, and the editor has the two tabs: Visual and Text, `triggerSave()` should not be called when the Text tab is "active". 

The reason is that if the user has edited something in Text, it will be overridden by the editor instance (the hidden Visual tab).

Best is to always use:
```js
var editor = tinymce.get( 'my-editor-id' );

if ( ! editor.isHidden() ) {
    editor.save(); // this also returns the cleaned up html
}

**Comment by webmandesign** [comment]
Thank you so much for looking at this ans providing compiled zip file for testing @youknowriad !

**Comment by youknowriad** [comment]
@azaozz Interesting, since we don't control the TinyMCE instances in Gutenberg, is there a way to retrieve all the instances in a given DOM container?

**Comment by azaozz** [comment]
TinyMCE keeps a collection of them, in `tinymce.editors`. Can perhaps look them up by (HTML) id?

**Comment by robincornett** [comment]
CMB2 has added a fix for this in their plugin now; [their solution](https://github.com/CMB2/CMB2/pull/1190) may be worth examining. I'm implementing it for my own work and it is working in 5.0RC without Gutenberg active.

**Comment by youknowriad** [comment]
@azaozz In my testing, even if I'm in the text mode, the fix in #12363 doesn't reset the value to be saved. I'd prefer to keep the fix simple. Would you mind reviewing the PR?

**Comment by webmandesign** [comment]
@youknowriad @azaozz I can confirm, the fix does not override anything during post saving even when in "Text" mode of TinyMCE.

I've tested with 2 different TinyMCE editors within 1 metabox, switching between modes, trying different combinations for both. Everything works perfectly fine.

Would be great to get this with WP5.0 ;)

**Comment by pglewis** [comment]
> 5.0 is frozen at the moment. This will more likely land in 5.0.1 which should happen just two weeks after the 5.0 but this could change, not my decision.

Do you know if this is still the plan?  I understand feature freeze, but this is a back-compat fix and seems like feature freeze should not stop it.  

If it does get punted to 5.0.1 what is the recommended course of action for plugin developers?  Sniff out 5.0 specifically and manually call a `triggerSave` hack-around, but not on 5.0.1+?

**Comment by pglewis** [comment]
Also: in preliminary testing the combination of the zip file above and the latest 5.0RC (but not 4.9.8) does resolve the problem for us with Pods Framework.  

Many Imaginary Internet Points for this fix landing in 5.0 because it makes one of my own issues evaporate :D.

**Comment by youknowriad** [comment]
closed by #12568
